### PR TITLE
Setup GitLab CI

### PR DESCRIPTION
**Pull request by benson31**
This adds testing pipelines for Catalyst, Corona, Lassen, Pascal, and Ray.

Due to LSF problems, Lassen and Ray test everything in one bulk job. Catalyst, Corona, and Pascal use Slurm and are able to break their testing into more precise stages/jobs. Developer documentation is hosted separately on Confluence. The ReadTheDocs documentation is somewhat minimal for now.

Testing the most recent changes is in progress. Feel free to check pipelines in the meantime, but I'll comment a final pipeline when I'm ready.

### CODE REVIEW

**Comment by benson31** [comment]
Here's a [pipeline](https://lc.llnl.gov/gitlab/benson31/tomslbannfork/-/pipelines/138987) for review.

**Review by timmoon10 — APPROVED** [review]
LGTM

**Review by mrwyattii — APPROVED** [review]

**Comment by benson31** [comment]
[New pipeline.](https://lc.llnl.gov/gitlab/benson31/tomslbannfork/-/pipelines/139265)

File: .gitlab/catalyst/pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    - export JOB_ID=$(squeue -h -n "${JOB_NAME}" -o "%A")
+    - ([[ -n "${JOB_ID}" ]] && scancel ${JOB_ID})
+
+# This loads the spack shell integration and then loads the
+# environment. Since this relies on the environment for this pipeline
+# existing, it should only be used after the build stage (i.e., in
+# testing).
+.uses spack environment:
+  before_script:
+    - source ${HOME}/spack_repos/spack_catalyst.git/share/spack/setup-env.sh
```
Comment (bvanessen) [inline .gitlab/catalyst/pipeline.yml:205]: To make this easier to copy and paste, we should probable replace catalyst here with ${SYSTEM_NAME}

File: .gitlab/corona/pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  when: always
+  script:
+    - echo "== RELEASING RESOURCES =="
+    - export JOB_ID=$(squeue -h -n "${JOB_NAME}" -o "%A")
+    - ([[ -n "${JOB_ID}" ]] && scancel ${JOB_ID})
+
+# Load the spack shell integration and load the environment.
+.uses spack environment:
+  before_script:
+    - source ${HOME}/spack_repos/spack_corona.git/share/spack/setup-env.sh
```
Comment (bvanessen) [inline .gitlab/corona/pipeline.yml:167]:
```suggestion
    - source ${HOME}/spack_repos/spack_${SYSTEM_NAME}.git/share/spack/setup-env.sh
```

File: .gitlab/catalyst/pipeline.yml (hunk 1 above)
Comment (bvanessen) [inline .gitlab/catalyst/pipeline.yml:205]:
```suggestion
    - source ${HOME}/spack_repos/spack_${SYSTEM_NAME}.git/share/spack/setup-env.sh
```

**Review by bvanessen — APPROVED** [review]
Works for me.

**Pull request merged**